### PR TITLE
[FIX] account: fix misc reverse tags

### PR DESCRIPTION
**Pull request by jbw-odoo**
Before this fix, all moves of type “entry“ were considered as misc operations in regard to repartition tags and the tax report. And only moves of type “out_refund” and “in_refund” were considered refunds.
Now reverse entries with sale/purchase taxes get their tags set accordingly to repartition lines.

```
Account                                    D      C      Tax   Tax Grids
700000 Ventes en Belgique (marchandises)   0      1000   21%   -03
451000 T.V.A. à payer                      0      210          -54
400000 Clients                             1210   0
```

Was reversed to:
```
Account                                    D      C      Tax   Tax Grids
700000 Ventes en Belgique (marchandises)   1000   0      21%   -03
451000 T.V.A. à payer                      210    0            -54
400000 Clients                             0      1210
```

Is now reversed to:
```
Account                                    D      C      Tax   Tax Grids
700000 Ventes en Belgique (marchandises)   1000   0      21%   +49
451000 T.V.A. à payer                      210    0            +64
400000 Clients                             0      1210

```
Task: 2687215

### CODE REVIEW

**Comment by jbw-odoo** [comment]
@oco-odoo

File: addons/account/models/account_move.py [hunk 1]
```diff
@@ -2096,8 +2116,7 @@ def compute_tax_repartition_lines_mapping(move_vals):
                 'debit': balance > 0.0 and balance or 0.0,
                 'credit': balance < 0.0 and -balance or 0.0,
             })
-
-            if move_vals['type'] not in ('out_refund', 'in_refund'):
+            if not is_refund:
                 continue
```
Comment (oco-odoo) [inline addons/account/models/account_move.py:2120]: This condition should come before the loop, directly returning, shouldn't it ? (already before your fix, but while we're at it ...)

If you move it, add a comment at the beginning of the loop indicating that it is only executed for refunds

File: addons/account/models/account_move.py [hunk 2]
```diff
@@ -3350,9 +3374,6 @@ def _compute_tax_audit(self):
             audit_str = ''
             for tag in record.tag_ids:
 
-                caba_origin_inv_type = record.move_id.type
-                caba_origin_inv_journal_type = record.journal_id.type
-
```
Comment (oco-odoo) [inline addons/account/models/account_move.py:3355]: Put this in another commit, so that we don't mix it with the content of the fix.

File: addons/account/tests/test_reconciliation.py [hunk 3]
```diff
@@ -1913,12 +1913,15 @@ def test_reconciliation_cash_basis_fx_02(self):
...
-        self.tax_waiting_account.reconcile = True
         tax_waiting_account10 = self.tax_waiting_account.copy({
             'name': 'TAX WAIT 10',
             'code': 'TWAIT1',
         })
+        tax_cash_basis10percent = self.tax_cash_basis.copy({
+            'amount': 10,
+            'cash_basis_transition_account_id': tax_waiting_account10.id,
+        })
+        self.tax_waiting_account.reconcile = True
```
Comment (oco-odoo) [inline addons/account/tests/test_reconciliation.py:1924]: Why this change ?

File: addons/account/models/account_move.py [hunk 4]
```diff
@@ -3378,7 +3399,11 @@ def _get_refund_tax_audit_condition(self, aml):
         whether or not it comes from a refund operation.
         This is overridden by pos in order to treat returns properly.
         """
-        return aml.move_id.type in ('in_refund', 'out_refund')
+        return (
+            aml.move_id.type in ('in_refund', 'out_refund') or
+            (aml.move_id.type == 'entry' and aml.move_id.journal_id.type == 'sale' and aml.debit > 0) or
+            (aml.move_id.type == 'entry' and aml.move_id.journal_id.type == 'purchase' and aml.credit > 0)
```
Comment (oco-odoo) [inline addons/account/models/account_move.py:3405]: I don't think this is right (and from my test: making a misc entry equivalent to an out refund gives the wrong tax audit, even though tags are properly inverted). We are already inverting the tags when needed, so I would say instead of doing this, we should do the same as in the tax report: here https://github.com/odoo/odoo/blob/13.0/addons/account/models/account_move.py#L3360 . We should also check things from POS there, so you'll probably need a hook doing the same as on your enterprise branch, overriding it in point_of_sale module.

**Review by oco-odoo — CHANGES_REQUESTED** [review]

File: addons/account/models/account_move.py (hunk 1 above)
Comment (jbw-odoo) [reply addons/account/models/account_move.py:2120]: line_vals are updated before this line, refund or not.

File: addons/account/tests/test_reconciliation.py (hunk 3 above)
Comment (jbw-odoo) [reply addons/account/tests/test_reconciliation.py:1924]: This is now considered as a refund, therefore revers lines have their account updated by _get_default_tax_account. The tax cash_basis_transition_account_id is now used instead of the original line account id.
As 'tax_cash_basis10percent' is a copy of 'tax_cash_basis' its needs it's cash_basis_transition_account_id updated. If not the reversed tax line wil get the tax_cash_basis cash_basis_transition_account_id instaed of  tax_waiting_account10 and reconciliation will not be possible.

**Comment by jbw-odoo** [comment]
Changes pushed

File: addons/account/models/account_move.py (hunk 1 above)
Comment (oco-odoo) [reply addons/account/models/account_move.py:2120]: I read it to quickly, indeed :eyeglasses:

File: addons/account/models/account_move.py [hunk 5]
```diff
@@ -3350,14 +3374,11 @@ def _compute_tax_audit(self):
...
 
-                caba_origin_inv_type = record.move_id.type
-                caba_origin_inv_journal_type = record.journal_id.type
-
                 if record.move_id.tax_cash_basis_rec_id:
                     # Cash basis entries are always treated as misc operations, applying the tag sign directly to the balance
                     type_multiplicator = 1
                 else:
-                    type_multiplicator = (record.journal_id.type == 'sale' and -1 or 1) * (self._get_refund_tax_audit_condition(record) and -1 or 1)
+                    type_multiplicator = (record.journal_id.type == 'sale' and self._get_not_entry_condition(record) and -1 or 1) * (self._get_refund_tax_audit_condition(record) and -1 or 1)
```
Comment (oco-odoo) [inline addons/account/models/account_move.py:3381]: Actually, this (and the _get_not_entry_condition function and override) are part of another fix than the reverse needing to create a refund, so we should make a distinct commit for it as well (and update PR description)

File: addons/account/models/account_move.py [hunk 6]
```diff
@@ -3373,6 +3394,13 @@ def _compute_tax_audit(self):
 
             record.tax_audit = audit_str
 
+    def _get_not_entry_condition(self, aml):
+        """
+        Returns the condition to exclude entry move types
```
Comment (oco-odoo) [inline addons/account/models/account_move.py:3399]: Please be a little clearer on the fact that this is used to compute tax_audit :p

**Review by oco-odoo — CHANGES_REQUESTED** [review]
Last details, nearly nothing :)

**Review by oco-odoo — APPROVED** [review]
@robodoo r+ rebase-ff

File: addons/account/tests/test_reconciliation.py (hunk 3 above)
Comment (smetl) [reply addons/account/tests/test_reconciliation.py:1924]: This is representing a bug in 13.0... the transition tax account is lost when reverting...
